### PR TITLE
Ref fields feature requires ByRefFields runtime flag

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/62919

Relates to test plan https://github.com/dotnet/roslyn/issues/59194

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/MissingCorLibrarySymbol.cs [hunk 1]
```diff
@@ -79,5 +79,20 @@ internal override bool RuntimeSupportsNumericIntPtr
                 throw ExceptionUtilities.Unreachable;
             }
         }
+
+        internal override bool RuntimeSupportsByRefFields
+        {
+            get
+            {
+                // For now we assume that it is not supported by default
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/MissingCorLibrarySymbol.cs:87]: For now? Is this ever going to change?

File: src/Compilers/CSharp/Portable/Symbols/MissingCorLibrarySymbol.cs [hunk 2]
```diff
@@ -79,5 +79,20 @@ internal override bool RuntimeSupportsNumericIntPtr
...
+                Debug.Assert((object)CorLibrary == this);
+                return false;
+            }
+            set
+            {
+                Debug.Assert((object)CorLibrary == this);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/MissingCorLibrarySymbol.cs:93]: Is this assert meaningful?

File: src/Compilers/VisualBasic/Test/Symbol/SymbolDisplay/SymbolDisplayTests.vb [hunk 3]
```diff
@@ -5283,7 +5283,11 @@ ref struct S<T>
     ref readonly T F2;
 }"
             Dim comp = CreateCSharpCompilation(GetUniqueName(), source, parseOptions:=New CSharp.CSharpParseOptions(CSharp.LanguageVersion.Preview))
-            comp.VerifyDiagnostics()
+            comp.VerifyDiagnostics(
```
Comment (333fred) [inline src/Compilers/VisualBasic/Test/Symbol/SymbolDisplay/SymbolDisplayTests.vb:5286]: Please use `AssertTheseDiagnostics`

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 1)

File: src/Compilers/CSharp/Portable/Symbols/MissingCorLibrarySymbol.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/MissingCorLibrarySymbol.cs:87]: Just copied previous property ;-)
To answer your question, this whole property will be removed once TargetingFramework.Net70 gets added, so we likely won't change the default.

File: src/Compilers/CSharp/Portable/Symbols/MissingCorLibrarySymbol.cs (hunk 2 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/MissingCorLibrarySymbol.cs:93]: Also still following the pattern established by Aleksey back in DIM: [https://github.com/dotnet/roslyn/pull/17927/files#diff-25f1d6b9209e1dd8d317520686107ee7a0eff64b4e4696570d8ff7eaa32079a9](https://github.com/dotnet/roslyn/pull/17927/files#diff-25f1d6b9209e1dd8d317520686107ee7a0eff64b4e4696570d8ff7eaa32079a9)

I would rather avoid variation unless there's a strong reason, as we're going to have to do this kind of test hook for many features moving forward.

File: src/Compilers/VisualBasic/Test/Symbol/SymbolDisplay/SymbolDisplayTests.vb (hunk 3 above)
Comment (jcouv) [reply src/Compilers/VisualBasic/Test/Symbol/SymbolDisplay/SymbolDisplayTests.vb:5286]: Had tried that, but it fails with `System.InvalidCastException : Unable to cast object of type 'Microsoft.CodeAnalysis.CSharp.CSharpCompilation' to type 'Microsoft.CodeAnalysis.VisualBasic.VisualBasicCompilation'.`

Comment (jcouv) [reply src/Compilers/VisualBasic/Test/Symbol/SymbolDisplay/SymbolDisplayTests.vb:5286]: Also, I'd considered getting ride of the diagnostic, by adding a runtime flag, but the C# test hook can't easily be reached from VB tests.
Another option is to skip the test until we have TargetFramework.Net70

Comment (333fred) [reply src/Compilers/VisualBasic/Test/Symbol/SymbolDisplay/SymbolDisplayTests.vb:5286]: Then please at least include what these errors are.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 4]
```diff
@@ -30,6 +31,26 @@ private static bool IsNet70OrGreater()
 
         private static string IncludeExpectedOutput(string expectedOutput) => IsNet70OrGreater() ? expectedOutput : null;
 
+        private static MetadataReference MscorlibRefWithoutSharingCachedSymbols
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:34]: >private static MetadataReference MscorlibRefWithoutSharingCachedSymbols [](http://example.com/codeflow?start=8&length=71)

Can this be a `static readonly` field with initializer so we don't need to copy the value to a local in methods that create more than one compilation?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 5]
```diff
@@ -90,7 +111,8 @@ void M5(S<T> s5)
         M2(ref s5.F1);
     }
 }";
-            var comp = CreateCompilation(sourceA, parseOptions: TestOptions.Regular10);
+            var mscorlibWithRefFields = MscorlibRefWithoutSharingCachedSymbols;
+            var comp = CreateByRefFieldsCompilation(sourceA, references: new[] { mscorlibWithRefFields }, parseOptions: TestOptions.Regular10);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:115]: >CreateByRefFieldsCompilation(sourceA, references: new[] { mscorlibWithRefFields }, parseOptions: TestOptions.Regular10) [](http://example.com/codeflow?start=23&length=119)

Could we hide most of the work in `CreateCompilation()`?

For instance, add a parameter to `CreateCompilation(..., RuntimeFeatures runtimeFeatures, ...)` that creates an cloned version of the appropriate corlib with the necessary features enabled, and that cloned version is created at most once and shared with all callers that need the same set of features.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 6]
```diff
@@ -166,15 +184,24 @@ public void RefField(bool useCompilationReference)
     public ref T F;
     public S(ref T t) { F = ref t; }
 }";
-            var comp = CreateCompilation(sourceA, parseOptions: TestOptions.Regular10);
+            // Avoid sharing mscorlib symbols with other tests since we are about to change
+            // RuntimeSupportsByRefFields property for it.
+            var mscorlibRefWithRefFields =
+                ((AssemblyMetadata)((MetadataImageReference)MscorlibRef).GetMetadata()).CopyWithoutSharingCachedSymbols().
+                     GetReference(display: "mscorlib.v4_0_30319.dll");
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:191]: Can we calculate this once and cache the value, perhaps in CSharpTestBase?

Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:191]: Or consider extracting a helper method if each caller needs a distinct instance.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 7]
```diff
@@ -166,15 +184,24 @@ public void RefField(bool useCompilationReference)
...
+
+            // Note: we use skipUsesIsNullable and skipExtraValidation so that nobody pulls
+            // on the compilation or its references before we set the RuntimeSupportsByRefFields flag.
+            var comp = CreateEmptyCompilation(sourceA, references: new[] { mscorlibRefWithRefFields }, parseOptions: TestOptions.Regular10, skipUsesIsNullable: true, skipExtraValidation: true);
+            comp.Assembly.RuntimeSupportsByRefFields = true;
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:196]: Can we use `CreateRefFieldsCompilation()` here and in other tests that create a copy of corlib?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 4 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs]: There's only two tests that use this. What are you trying to achieve?

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 4 above)
Comment (cston) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs]: I was mistaken. I thought this was used in more tests, and that all tests could share a common reference.

**Comment by jcouv** [comment]
@333fred for second review. Thanks

**Comment by jcouv** [comment]
@333fred for second review. Thanks

File: src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs [hunk 8]
```diff
@@ -467,8 +467,19 @@ protected bool RuntimeSupportsFeature(SpecialMember feature)
         internal bool RuntimeSupportsUnmanagedSignatureCallingConvention
             => RuntimeSupportsFeature(SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__UnmanagedSignatureCallingConvention);
 
-        internal bool RuntimeSupportsByRefFields
-            => RuntimeSupportsFeature(SpecialMember.System_Runtime_CompilerServices_RuntimeFeature__ByRefFields);
+        internal virtual bool RuntimeSupportsByRefFields
+        {
+            get
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/AssemblySymbol.cs:472]: Nit: please use consistent style between these two one-line methods :).

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 9]
```diff
@@ -90,7 +99,12 @@ void M5(S<T> s5)
         M2(ref s5.F1);
     }
 }";
-            var comp = CreateCompilation(sourceA, parseOptions: TestOptions.Regular10);
+            var mscorlibRefWithRefFields = GetMscorlibRefWithoutSharingCachedSymbols();
+
+            // Note: we use skipUsesIsNullable and skipExtraValidation so that nobody pulls
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:104]: I'm confused why this is necessary. I wouldn't expect anything to be pulling on the compilation or it's references until the compilation is actually used? Can you clarify why we need to do this for this particular flag, but haven't needed to for any of the other `RuntimeSupports` bypass flags?

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 10]
```diff
@@ -30,6 +31,14 @@ private static bool IsNet70OrGreater()
 
         private static string IncludeExpectedOutput(string expectedOutput) => IsNet70OrGreater() ? expectedOutput : null;
 
+        private static IEnumerable<MetadataReference> CopyWithoutSharingCachedSymbols(IEnumerable<MetadataReference> references)
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:34]: Why is this necessary? It wasn't necessary for any previous feature that set override flags.

**Review by 333fred — COMMENTED** [review]
Done review pass (commit 9). @jcouv and I need to sync on the approach taken in the tests, as no previous bypass flag has needed this level of modification before.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 9 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:104]: We don't need this for this particular flag, but we need it for certain flags such as numeric IntPtr (pattern that I'm following). 
I'll remove but we should have a discussion at some point on a universal test hook, so that we don't have to special-case every new runtime feature flag.

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:34]: This is used in two tests that need more than mscorlib. For example, one needs `dynamic` machinery.
Unless those previous features tests such combinations, they would not run into this situation.
I can't say that I understand `CopyWithoutSharingCachedSymbols` well enough to say there isn't a better way to do this. I'm happy to open a tracking issue to re-discuss when Aleksey is back.

**Review by cston — APPROVED** [review]

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 10 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:34]: Filed [https://github.com/dotnet/roslyn/issues/63222](https://github.com/dotnet/roslyn/issues/63222) to review usages of `CopyWithoutSharingCachedSymbols()`

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs (hunk 9 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs]: I opened an issue ([https://github.com/dotnet/roslyn/issues/63223](https://github.com/dotnet/roslyn/issues/63223)) to sit down and design a general/re-usable CreateCompilation API supporting what we need, including a note on `skipUsesIsNullable` and `skipExtraValidation` since some features need those but some don't.
